File: baekjoon/dynamic_programming/11055_가장_큰_증가_부분_수열.cpp
#include <cstdio>

int n, ans, arr[1001], dp[1001][2];

int getMax(int a, int b) { return a > b ? a : b; }

int main(void) {
	scanf("%d", &n);

	for (int i = 0; i < n; i++) {
		scanf("%d", &arr[i]);
		dp[i][0] = 1;
		dp[i][1] = arr[i];
	}

	ans = dp[0][1];

	for (int i = 0; i < n; i++) {
		for (int j = 0; j < i; j++) {
			if (arr[i] > arr[j] && dp[j][0] + 1 > dp[i][0]) {
				dp[i][0] = dp[j][0] + 1;
				dp[i][1] = getMax(dp[i][1], dp[j][1] + arr[i]);
			}

			ans = getMax(ans, dp[i][1]);
		}
	}

	printf("%d", ans);
		
	return 0;
}

File: baekjoon/dynamic_programming/11055_가장큰증가부분수열.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {
  public static void main(String[] args) throws IOException {
    int ans = 0;
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    int n = Integer.parseInt(br.readLine());
    int[] arr = new int[n];
    int[] dp = new int[n];
    String[] str = br.readLine().split(" ");

    for (int i = 0; i < str.length; i++) {
      arr[i] = Integer.parseInt(str[i]);
    }

    for (int i = 0; i < n; i++) {
      dp[i] = arr[i];
      int max = 0;
      for (int j = 0; j < i; j++) {
        if (arr[i] > arr[j]) {
          max = Math.max(max, dp[j]);
        }
      }

      dp[i] += max;
      ans = Math.max(ans, dp[i]);
    }

    System.out.println(ans);
  }
}
